README.md
# NAME

Nephia::Plugin::Dispatch - Dispatcher Plugin for Nephia

# DESCRIPTION

This plugin provides dispatcher feature to Nephia.

# SYNOPSIS

    use Nephia plugins => ['Dispatch'];
    

    my $users = {};
    

    app {
        get '/' => sub { [200, [], 'Hello, World!'] };
        get '/user/:id' => sub {
            my $id = path_param('id');
            my $user = $users->{$id};
            $user ? 
                [200, [], sprintf('name = %s', $user->{name}) ]
                [404, [], 'no such user']
            ;
        };
        post '/user/:id' => sub {
            my $id = path_param('id');
            my $name = param('name'); 
            $users->{$id} = { name => $name };
            [200, [], 'registered!'];
        };
    };

# DSL

## get post put del

    get $path => sub { ... };

Add action for $path. You may use [Router::Simple](http://search.cpan.org/perldoc?Router::Simple) syntax in $path.

## path\_param

    get '/user/:id' => sub {
        my $id = path_param('id');
        ### or 
        my $path_params = path_param;
        $id = $path_params->{id};
        ...
    };

Fetch captured parameter from PATH\_INFO.

# LICENSE

Copyright (C) ytnobody.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

ytnobody <ytnobody@gmail.com>


Changes
Revision history for Perl extension Nephia-Plugin-Dispatch

{{$NEXT}}

0.03 2013-11-30T07:35:40Z
    - Using res_404 when it available

0.02 2013-10-26T02:23:41Z
    - change prototype interface for get/post/put/del

0.01 2013-09-28T23:22:44Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Voson-Plugin-Dispatch',
    module_name     => 'Voson::Plugin::Dispatch',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Nephia/Plugin/Dispatch.pm
package Nephia::Plugin::Dispatch;
use 5.008005;
use strict;
use warnings;
use parent 'Nephia::Plugin';
use Router::Simple;

our $VERSION = "0.02";

sub exports {
    qw/get post put del path_param/;
}

sub new {
    my ($class, %opts) = @_;
    my $self = $class->SUPER::new(%opts);
    my $app = $self->app;
    $app->action_chain->after('Core', Dispatch => $self->can('dispatch'));
    $app->{router} = Router::Simple->new;
    return $self;
}

sub dispatch {
    my ($app, $context) = @_;
    my $router = $app->{router};
    my $req    = $context->get('req');
    my $env    = $req->env;
    my $res    = $app->dsl('res_404') ? $app->dsl('res_404')->() : [404, [], ['not found']];
    if (my $p = $router->match($env) ) {
        my $action = delete $p->{action};
        $context->set(path_param => $p);
        $res = $action->($app, $context);
    }
    $context->set(res => $res);
    return $context;
}

sub path_param {
    my ($self, $context) = @_;
    return sub (;$) {
        my $path_param = $context->get('path_param');
        $_[0] ? $path_param->{$_[0]} : $path_param;
    };
}

sub path {
    my ($self, $context, $method) = @_;
    my $router = $self->app->{router};
    return sub ($;@) {
        my ($path, $code) = @_;
        my @pathes = ref($path) eq 'ARRAY' ? @$path : ( $path );
        $router->connect($_, {action => $code}, {method => $method}) for @pathes;
    };
}

{
    no strict qw/refs/;
    my %methods = (get => 'GET', post => 'POST', put => 'PUT', del => 'DELETE');
    for my $dsl (keys %methods) {
        *$dsl = sub {
            my ($self, $context) = @_;
            $self->path($context, $methods{$dsl});
        };
    };
}

1;

__END__

=encoding utf-8

=head1 NAME

Nephia::Plugin::Dispatch - Dispatcher Plugin for Nephia

=head1 DESCRIPTION

This plugin provides dispatcher feature to Nephia.

=head1 SYNOPSIS

    package My::NephiaApp;
    use Nephia plugins => ['Dispatch'];
    
    {   ### External Controller Class
        package My::NephiaApp::C::External;
        sub index {
            my $c = shift;             # Nephia::Core object
            my $id = $c->param('id');  # You may call param method via Nephia::Core object
            [200, [], ["id = $id"]];
        }
    };

    my $users = {};
    
    app {
        get '/' => sub { [200, [], 'Hello, World!'] };
        get '/user/:id' => sub {
            my $id = path_param('id');
            my $user = $users->{$id};
            $user ? 
                [200, [], sprintf('name = %s', $user->{name}) ]
                [404, [], 'no such user']
            ;
        };
        post '/user/:id' => sub {
            my $id = path_param('id');
            my $name = param('name'); 
            $users->{$id} = { name => $name };
            [200, [], 'registered!'];
        };
        get '/external/' => Nephia->call('C::External#index');
    };
    

=head1 DSL

=head2 get post put del

    get $path => sub { ... };

Add action for $path. You may use L<Router::Simple> syntax in $path.

=head2 path_param

    get '/user/:id' => sub {
        my $id = path_param('id');
        ### or 
        my $path_params = path_param;
        $id = $path_params->{id};
        ...
    };

Fetch captured parameter from PATH_INFO.

=head1 LICENSE

Copyright (C) ytnobody.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

ytnobody E<lt>ytnobody@gmail.comE<gt>

=cut

